src/matrixrtc/CallMembership.ts

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2023 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { MatrixEvent, RoomMember } from "../matrix";
+import { deepCompare } from "../utils";
+import { Focus } from "./focus";
+
+type CallScope = "m.room" | "m.user";
+
+// Represents an entry in the memberships section of an m.call.member event as it is on the wire
+export interface CallMembershipData {
+    application?: string;
+    call_id: string;
+    scope: CallScope;
+    device_id: string;
+    created_ts?: number;
+    expires: number;
+    foci_active?: Focus[];
+}
+
+export class CallMembership {
+    public static equal(a: CallMembership, b: CallMembership): boolean {
+        return deepCompare(a.data, b.data);
+    }
+
+    public constructor(private parentEvent: MatrixEvent, private data: CallMembershipData) {
+        if (typeof data.expires !== "number") throw new Error("Malformed membership: expires must be numeric");
+        if (typeof data.device_id !== "string") throw new Error("Malformed membership event: device_id must be string");
+        if (typeof data.call_id !== "string") throw new Error("Malformed membership event: call_id must be string");
+        if (typeof data.scope !== "string") throw new Error("Malformed membership event: scope must be string");
```

Comment:
Perhaps a dumb question from my side: Aren't the fields typed? - I see the types defined? (We're using TypeScript, right?) Or is it because currently, we allow type mismatches on the SDK level (no strictness?)

Reply:
So this is from where we're doing `getContent()` (line 73) and declaring that variable as a `CallMembershipData[]`. We don't actually know what is in the event: it could be anything, so we do need to validate it at some point. I went back & forth a few times on how to do this: we could have this just accept a `Record<string, unknown>` which is more like what the event data actually is and may be considered more correct, but it felt more helpful to define the right shape & then validate it.